lib/spack/spack/package.py

```diff
@@ -1612,6 +1612,10 @@ def run_test(self, exe, options=[], expected=[], status=None,
                 if runner is None and skip_missing:
                     return
 
+                if skip_missing and installed and \
+                        self.spec.prefix not in runner.path:
+                    return
+
```

Comment:
This is exactly the opposite of the intention of the `installed` parameter. The whole point of it is to fail if the package didn't install some particular executable.

For hdf5, let's just skip testing any executable that we can't write a simple `if spec.satisfies(___)` to express whether it should be there.

Reply:
The problem I was trying to address occurs if you have an `hdf5` binary in your path that is **not** installed.

Reply:
> This is exactly the opposite of the intention of the `installed` parameter. The whole point of it is to fail if the package didn't install some particular executable.

Can we discuss this?  

As I mention in my previous response, I was getting failures when the binaries were being acquired from my path (on LC) instead of the installation directory; hence, my change.

> 
> For hdf5, let's just skip testing any executable that we can't write a simple `if spec.satisfies(___)` to express whether it should be there.

Can we discuss this as well?

As I've found with other packages, this can be a challenge should developers decide to remove then restore or move the executables (e.g., `umpire` #17178 ).

Reply:
Re: PATH ordering 

The testing environment should have put the binaries from the HDF5 we're testing before anything else in the PATH. If we're picking up a different hdf5 at a higher priority, I think that's an error in how we're setting up the run environment for hdf5.